File: 13-05-2024/Number of Good Components.py

from typing import List

#Your Code Start
class Solution:
    def isConnected(self, check: int, visited: List[bool], adj: List[List[int]]) -> bool:
        check_list = adj[check][:]
        check_list.append(check)
        check_list.sort()
        
        for i in adj[check]:
            visited[i] = True
            temp = adj[i][:]
            temp.append(i)
            temp.sort()
            if check_list != temp:
                return False
        return True
    
    def findNumberOfGoodComponent(self, e: int, v: int, edges: List[List[int]]) -> int:
        ans = 0
        visited = [False] * (v + 1)
        adj = [[] for _ in range(v + 1)]
        
        for edge in edges:
            start, end = edge
            adj[start].append(end)
            adj[end].append(start)
        
        for i in range(1, v + 1):
            if not visited[i]:
                isConnectedComponent = self.isConnected(i, visited, adj)
                if isConnectedComponent:
                    ans += 1
        
        return ans
#Your Code End

#{ 
 # Driver Code Starts
class IntMatrix:

    def __init__(self) -> None:
        pass

    def Input(self, n, m):
        matrix = []
        #matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix

    def Print(self, arr):
        for i in arr:
            for j in i:
                print(j, end=" ")
            print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):

        e = int(input())

        v = int(input())

        edges = IntMatrix().Input(e, 2)

        obj = Solution()
        res = obj.findNumberOfGoodComponent(e, v, edges)

        print(res)

# } Driver Code Ends

File: 13-05-2024/Number of Good Components.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Matrix {
  public:
    template <class T>
    static void input(vector<vector<T>> &A, int n, int m) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                scanf("%d ", &A[i][j]);
            }
        }
    }

    template <class T>
    static void print(vector<vector<T>> &A) {
        for (int i = 0; i < A.size(); i++) {
            for (int j = 0; j < A[i].size(); j++) {
                cout << A[i][j] << " ";
            }
            cout << endl;
        }
    }
};


// } Driver Code Ends

//Your Code Start
class Solution {
public:
    bool isConnected(int check, vector<bool>& visited, vector<vector<int>>& adj) {
        vector<int> checkVector = adj[check];
        checkVector.push_back(check);
        sort(checkVector.begin(), checkVector.end());
        
        for(auto i : adj[check]) {
            visited[i] = true;
            vector<int> temp = adj[i];
            temp.push_back(i);
            sort(temp.begin(), temp.end());
            if(checkVector != temp) return false;
        }
        return true;
    }
    
    int findNumberOfGoodComponent(int e, int v, vector<vector<int>> &edges) {
        int ans = 0;
        vector<bool> visited(v + 1, false);
        vector<vector<int>> adj(v + 1, vector<int>());
        
        for(auto edge : edges) {
            int start = edge[0];
            int end = edge[1];
            
            adj[start].push_back(end);
            adj[end].push_back(start);
        }
        
        for(int i = 1; i <= v; i++) {
            if(!visited[i]) {
                bool isConnectedComponent = isConnected(i, visited, adj);
                if(isConnectedComponent) ans++;
            } 
        }
        
        return ans;
    }
};
//Your Code End

//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d ", &t);
    while (t--) {

        int e;
        scanf("%d", &e);

        int v;
        scanf("%d", &v);

        vector<vector<int>> edges(e, vector<int>(2));
        Matrix::input(edges, e, 2);

        Solution obj;
        int res = obj.findNumberOfGoodComponent(e, v, edges);

        cout << res << endl;
    }
}

// } Driver Code Ends
